Improvements to Docker file

dwd opened the pull request:
This changes the Dockerfile quite radically, so that Openfire is built within Docker rather than outside of it. This should simplify building the image, and also make the results more repeatable.

In order to maximize the use of the cache while keeping the image size under control, this uses three stages:

* The first stage locates and extracts all the POM files and any JAR files. This stage will be re-run on any change, but it short.
* The second stage takes the output of the first stage and gathers dependencies. These will be cached, unless the output of the first stage (ie, the dependency information) changes. Then it copies in the full source, and builds it.
* Finally, the runtime container is setup much as it was before, except that the runtime files are copied from the build stage rather than the filesystem directly.

The result is that a repeat build of the docker image now takes about two minutes, but can trivially be done on any docker platform (even without Java installed locally).

Notes:
* The build stage should be able to run the `mvn package` in offline mode, but maven (being maven) wants to download more during this stage.
* The `.dockerignore` file has of course been changed, but someone who understands Java better than I might well improve it further.

guusdk commented:
Thanks for this! I'm not Docker-savvy enough to review this properly. @Fishbowler @Fank can you have a look please?

Assuming that this builds Openfire from source, it may be able to leverage the Maven wrapper that's part of the repository (`./mvnw`) rather than download/install one.

Fishbowler commented on Dockerfile:
```diff
@@ -1,3 +1,35 @@
+FROM openjdk:11-jdk AS build
+# Set up Maven. No need to clean caches, this doesn't end up in the runtime container
+RUN apt-get update -qq
+RUN apt-get install -qqy maven
```
Agree with Guus' point - is there a good reason not to use the `./mvnw` wrapper used instead? I don't think it'd affect caching, right?

Fishbowler reviewed (changes requested):
Other than the one thing on Maven, this looks really good! 

First run: 534.9s
Second run: 0.9s :tada:

Fishbowler commented on Dockerfile:
```diff
@@ -1,3 +1,35 @@
+# This stage extracts all the pom.xml files.
+# It'll get rebuilt with any source change, but that's OK.
+FROM openjdk:11-jdk AS poms
```
openjdk is deprecated (https://hub.docker.com/_/openjdk)

Is `eclipse-temurin:11-jdk-noble` the closest similar?

Fishbowler commented:
Oooh, another experiential one.
When the JDK base image updates, it invalidates the docker cache at the first line.
Wonder if we could pin stuff for the early steps, then use latest for the later stuff? Or better to pin all the way down?

dwd commented:
> Oooh, another experiential one. When the JDK base image updates, it invalidates the docker cache at the first line. Wonder if we could pin stuff for the early steps, then use latest for the later stuff? Or better to pin all the way down?

I wouldn't think you'd want to try and avoid the hit there - once you start running Java, I think you want the latest possible. You could use something else for the extraction stage, but that would end up dominated by fetching the container anyway, which is why I chose to use the same one throughout.

dwd commented:
> Other than the one thing on Maven, this looks really good!
> 
> First run: 534.9s Second run: 0.9s 🎉

That looks like full cache. The better test is an arbitrary change (comment will do) in a source file, which is more typical developer experience.

I imagine we could make things faster by examining which packages are changed most frequently, and see if we could cache the build for some fo the less frequently changed ones, but that seems quite complex.

dwd commented:
Summary of those additional commits:
* Merged main
* Switch to our mvnw
* Used eclipse-temurin:17 as the base/JDK image
* Improved caching of dependencies
* Used a skeleton stage to improve readability/image size/parallelization
* Added docker build to workflow

TODO list, all can/should be done post-merge:
* If we added suitable secrets to GH, then the built image can be conditionally pushed to Docker Hub (or elsewhere)
* Modern method for Java docker images seems to be `jlink`, but this needs source support I think.

guusdk commented:
> If we added suitable secrets to GH, then the built image can be conditionally pushed to Docker Hub (or elsewhere)

We could consider publishing built images via the GitHub Packages system, which can act as a Docker registry. (That doesn't seem to require additional secrets). We've experimented recently with that in this project: https://github.com/XMPP-Interop-Testing/smack-sint-server-extensions/blob/main/.github/workflows/docker.yml

Fishbowler commented:
The more recent changes introduced a new problem:

```
> docker run --rm -it openfire:latest -demoboot
Initializing /var/lib/openfire...
chown: invalid group: ‘openfire:openfire’
```

Fishbowler commented:
I've popped on another commit for you to look at - it's got the groups, and moves the sudo install down to the last container, as its needed for the entrypoint. I don't think there's a way to move this any earlier, unless we want to layer it?

Fishbowler commented on .github/workflows/continuous-integration-workflow.yml:
```diff
@@ -139,14 +139,21 @@ jobs:
-            echo "is_publishable_branch=true" >> $GITHUB_OUTPUT
+          if [[ ${{ github.ref }} == 'refs/heads/main' ]]; then
+            echo "is_publishable_branch=true" >> "${GITHUB_OUTPUT}"
+            echo "branch_tag=latest" >> "${GITHUB_OUTPUT}"
```
Would we want `latest` pointing at the tip of main? We consider it unstable.
Would we be better having main be `main` or `bleeding_edge` or `unstable` or `testing` or something, and work out how to match the latest release with `latest`?

dwd replied:
It's now "development".

Fishbowler commented:
@stokito ` apt install openfire` assumes that you only wish to install a version of Openfire that exists in a public repository. I feel that might not work for the number of forks that exist, let alone individual developer workflows. I for one regularly test PRs in containers.

@dwd Do you have time to cast some stink eye over the mess I've made of your PR, and of my comments?

dwd commented:
> Assuming that this builds Openfire from source, it may be able to leverage the Maven wrapper that's part of the repository (`./mvnw`) rather than download/install one.

Done!

dwd commented:
> @dwd Do you have time to cast some stink eye over the mess I've made of your PR, and of my comments?

Also done, removed the need for sudo entirely now.

dwd commented:
> The more recent changes introduced a new problem:
> 
> ```
> > docker run --rm -it openfire:latest -demoboot
> Initializing /var/lib/openfire...
> chown: invalid group: ‘openfire:openfire’
> ```

I believe this now works.

dwd commented:
I'll resolve the conflicts shortly.

guusdk commented on mvnw:
```diff
@@ -50,7 +50,7 @@ fi
 darwin=false;
 mingw=false
-case "`uname`" in
+case "$(uname)" in
```
I was going to suggest to push this change upstream, but I believe that it's already applied there: https://github.com/apache/maven-wrapper/blame/maven-wrapper-3.3.2/maven-wrapper-distribution/src/resources/mvnw

Should we update our bundled Maven Wrapper (rather than manually modifying it)?

dwd replied:
Probably, though that feels like a distinct PR.

guusdk replied:
Ack: https://github.com/igniterealtime/Openfire/pull/2593

Fishbowler commented:
Works on my machine :heart:
Time to change that maintainer label?

guusdk commented:
> Time to change that maintainer label?

Given that @Fank seems to have moved on to greener pastures, and most of the original content now has been replaced, that seems reasonable.

Fishbowler reviewed (approved):
Let's change the maintainer to dwd and merge 🎉

guusdk commented:
I've rebased this, and switched out the maintainer label (now references @dwd and @Fishbowler ).

guusdk commented:
This change is now merged. Thanks for all the hard work that has gone into this!

Now that it is merged, should we have a look at automatically publishing a Docker and/or OCI image? The existing GitHub workflows contain a step that conditionally publish to DockerHub, but I think that's never used (and if so, I wouldn't know what DockerHub repository is ours). Should we re-instate that? Should we instead use the GitHub Release support for this (and distribute the images through that rather than through DockerHub)?

What versioning do we use? One image for every tag, and images for every branch (Main, vX.Y) that is merged to?

Fishbowler commented:
I used it for my own fork. It's never been used in Ignite.

guusdk commented:
I've copy/pasted most of @Fishbowler's work in xmpp-interop-test to replace what pre-existed with an attempt to publish to the GitHub Registry in https://github.com/igniterealtime/Openfire/pull/2642
